lib/web_console/request.rb

```diff
@@ -32,6 +32,49 @@ def acceptable_content_type?
       content_type.blank? || content_type.in?(acceptable_content_types)
     end
 
+    # Returns whether the request includes "application/vnd.web-console"
+    def has_media_type?
+      accept.split(/\s*;\s*/, 2)[0].split(/\s*,\s*/).any? do |t|
+        /^application\/vnd.web-console$/ === t
+      end
+    end
+
+    # Returns whether the request requires versioning in accept header
+    def has_version_requirement?
+      ! required_version.nil?
+    end
+
+    # Returns whether the request is acceptable
+    #
+    # It returns true if required version and current version have
+    # the same major and minor version.
+    # It also returns true if a request includes no vendor media type.
+    def acceptable_required_version?
```

Comment:
If we utilize the `Mime` module, we won't need to do all this parsing by ourselves. See the implementation of [`ActionDispatch::Request#accepts`](https://github.com/rails/rails/blob/261a8e1c4bcd08a7946a8ff8b59ba1085e4aab1e/actionpack/lib/action_dispatch/http/mime_negotiation.rb#L31-L41), if we register our own MIME Type trough it, we get parsing for free.

```ruby
>> Mime::Type.register 'application/vnd.web-console.v2', :web_console_v2
=> []
>> Mime::Type.parse('application/vnd.web-console.v2').first
=> #<Mime::Type:0x007fe8e9e35a88 @synonyms=[], @symbol=:web_console_v2, @string="application/vnd.web-console.v1">
>> Mime::Type.parse('application/vnd.web-console.v2').first == Mime::WEB_CONSOLE_V2
=> true
```

That way, we don't need `has_version_requirement?`, `has_media_type?`, `required_version` and the `gem_version` methods. Sadly, we would need to drop the parameter in the header, because while we can register a media type with a parameter, when we try to parse it back, the parameters will be trimmed. This is an offtopic, but its an area in which we can work on Rails upstream to make this support better.

```ruby
>> Mime::Type.register 'application/vnd.web-console; version=2', :web_console_v2
=> []
>> Mime::Type.parse('application/vnd.web-console; version=2')
=> [#<Mime::Type:0x007fcd3280fcd0 @synonyms=[], @symbol=nil, @string="application/vnd.web-console">]
>> Mime::Type.parse('application/vnd.web-console.v2').first == Mime::WEB_CONSOLE_V2
=> false
```

---

If we are to go with `Accept` headers like `application/vnd.web-console.v2`, which are still acceptable IMO, the only parsing method we'll need will look something like this:

```ruby
def acceptable_version?
  accepts.any? { |mime| mime == Mime::WEB_CONSOLE_V2 }
end
```

On the other side, I think that we don't need to follow Web Console versioning strictly. If we do, every time we release a minor version of the gem, we would have to update the extensions. I think it would be easier for us if we can release those separately. Say, we fixed a bug in the GEM, now the users have to update their extensions, because we needed to change a header.

We have to think about JSON API stability once we release the first extensions and keep it for a while. Who knows, if we don't have that many JSON API changes for Web Console v3, we may even support older versions of the extensions that still send `application/vnd.web-console.v2`.